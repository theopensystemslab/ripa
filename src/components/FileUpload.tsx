import Box from "@material-ui/core/Box";
import Button from "@material-ui/core/Button";
import makeStyles from "@material-ui/core/styles/makeStyles";
import Typography from "@material-ui/core/Typography";
import { useFormik } from "formik";
import React, { useCallback, useEffect, useState } from "react";
import { useDropzone } from "react-dropzone";

import Messages from "../shared/components/submit-messages";

const styles = theme => ({
  box: {
    width: 200,
    height: 200,
    background: "#EAEAEA",
    padding: theme.spacing(2),
    display: "flex",
    flexDirection: "column",
    justifyContent: "center"
  },
  button: {
    textDecoration: "underline",
    display: "inline",
    fontFamily: "inherit",
    border: 0,
    paddingLeft: 0,
    background: "none",
    cursor: "pointer",
    textAlign: "left"
  }
});
const useStyles = makeStyles(styles as any) as any;

interface IFileUpload {
  title: string;
  accept: string[];
  maxSize: number;
  includeSubmit?: boolean;
}

export const FileUpload: React.FC<IFileUpload> = ({
  title,
  maxSize,
  accept = [],
  includeSubmit = false
}) => {
  const [files, setFiles] = useState([]);
  const [stateText, setStateText] = useState("Click to select files");

  const classes = useStyles();
  const [errorMessageVisible, setErrorMessageVisible] = useState(false);

  const formik = useFormik({
    initialValues: {
      path: ""
    },
    onSubmit: values => {
      alert(JSON.stringify(values, null, 2));
    }
  });

  const onDrop = useCallback(
    acceptedFiles => {
      setErrorMessageVisible(false);
      console.log(acceptedFiles);
      if (acceptedFiles.length > 0) {
        const reader = new FileReader();
        reader.onprogress = () => {
          setStateText("Loading");
        };
        reader.onload = () => {
          setStateText("click to select files");
        };
        setFiles(acceptedFiles);
      } else {
        setErrorMessageVisible(true);
      }
    },
    [maxSize]
  );

  useEffect(() => {
    if (files.length > 0) {
      formik.setFieldValue("path", files[0].path);
    }
  }, [files]);

  const { getRootProps, getInputProps } = useDropzone({
    onDrop,
    accept,
    maxSize
  });
  return (
    <Box py={4}>
      <form onSubmit={formik.handleSubmit}>
        <Typography variant="h5" component="h1" gutterBottom>
          {title}
        </Typography>
        {files.map(file => {
          return (
            <div key={file.name} className={classes.box}>
              <p>
                Name: <strong>{file.name}</strong>
              </p>
              <p>Size: {file.size}</p>
              <p>Type: {file.type}</p>
              <p onClick={e => setFiles(files.filter(f => f !== file))}>x</p>
            </div>
          );
        })}
        <div className={classes.box} {...getRootProps()}>
          <input data-testid="fileupload" {...getInputProps()} />
          <p>{stateText}</p>
        </div>
        <small>Max size of file is {maxSize} Bytes</small>
        <div>
          {errorMessageVisible && formik.touched ? (
            <Messages
              type="error"
              message={`Please choose a file with max size ${maxSize} Bytes`}
            />
          ) : null}
        </div>
        {includeSubmit && (
          <div>
            <Button type="submit" variant="contained" color="primary">
              Save and Continue
            </Button>
          </div>
        )}
      </form>
    </Box>
  );
};

export default FileUpload;
