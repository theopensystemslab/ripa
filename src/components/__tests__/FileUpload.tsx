import { cleanup, fireEvent, render, wait } from "@testing-library/react";
import * as React from "react";

import FileUpload from "../FileUpload";

afterEach(cleanup);

describe("File Upload Component", () => {
  it("should render snapshot", () => {
    const maxSize = 40000;
    const accept = ["image/*"];
    const title = "File upload";
    const { asFragment } = render(
      <FileUpload maxSize={maxSize} accept={accept} title={title} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
