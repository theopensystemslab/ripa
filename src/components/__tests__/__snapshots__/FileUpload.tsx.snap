// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Upload Component should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-4"
  >
    <form>
      <h1
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
      >
        File upload
      </h1>
      <div
        class="makeStyles-box-1"
        tabindex="0"
      >
        <input
          accept="image/*"
          autocomplete="off"
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <p>
          Click to select files
        </p>
      </div>
      <small>
        Max size of file is 40000 Bytes
      </small>
      <div />
    </form>
  </div>
</DocumentFragment>
`;
